lib/BreakpointPartition.js

```diff
@@ -0,0 +1,62 @@
+export default function BreakpointPartition(imageRatioSequence, expectedRowCount) {
+  if (imageRatioSequence.length <= 1) return [imageRatioSequence];
+  if (expectedRowCount >= imageRatioSequence.length) return imageRatioSequence.map(item => [item]);
+
+  const layoutWidth = findLayoutWidth(imageRatioSequence, expectedRowCount);
+  let currentRow = 0;
+
+  return imageRatioSequence.reduce((rows, imageRatio) => {
+    if (sum(rows[currentRow]) + imageRatio > layoutWidth) currentRow++;
+    rows[currentRow].push(imageRatio);
+    return rows;
+    // waiting for more elegant solutions (Array.fill) to work correctly
+  }, new Array(expectedRowCount).join().split(',').map(() => []));
+}
+
+// starting at the ideal width, expand to the next breakpoint until we find
+// a width that produces the expected number of rows
+function findLayoutWidth(imageRatioSequence, expectedRowCount) {
+  let idealWidth = sum(imageRatioSequence) / expectedRowCount
+  let widestItem = max(imageRatioSequence);
+  let galleryWidth = max([idealWidth, widestItem]);
+  let layout = getLayoutDetails(imageRatioSequence, layoutWidth);
+
+  while (layout.rowCount > rowCount) {
+    layoutWidth += layout.nextBreakpoint
+
+    layout = getLayoutDetails(imageRatioSequence, layoutWidth);
+  }
+  return layoutWidth;
+}
+
+// find the
+function getLayoutDetails(imageRatioSequence, expectedWidth) {
+  const startingLayout = {
+    currentRowWidth: 0,
+    rowCount: 1,
+    // the largest possible step to the next breakpoint is the smallest image ratio
+    nextBreakpoint: Math.min.apply(null, imageRatioSequence)
+  }
+  const finalLayout = imageRatioSequence.reduce((layout, itemWidth) => {
+    const rowWidth = layout.currentRowWidth + itemWidth;
+    let nextBreakpoint;
+    if (rowWidth > expectedWidth) {
+      nextBreakpointForCurrentRow = layout.currentRowWidth - expectedWidth;
```

Comment:
Here is the key difference in this algorithm. When an image is too wide to fit on the current row, I'm calculating how much more width would be needed to include it. If it requires less space than any previous row needed, I track it. When we've passed through all rows, if the current layout produces too many rows, I make another pass at the smallest width that will allow an image to move up a row.